### ISSUE TITLE
Issues downloading mods if their jar files has spaces in the filename

### ISSUE THREAD

**Issue by mythmon** [Nincraft/ModPackDownloader#102]
The log and manifest used can be found [here](https://gist.github.com/mythmon/589ea705768b2b004ca856ee321ca6d0), though I expect the excerpt I posted here would be enough.

When downloading mods, I got this error: 

```
java.io.FileNotFoundException: Source '/root/.modpackdownloader/inventory-tweaks/InventoryTweaks-1.64 dev.151.jar' does not exist
        at org.apache.commons.io.FileUtils.checkFileRequirements(FileUtils.java:274) ~[ModpackDownloader.jar:?]
        at org.apache.commons.io.FileUtils.copyFile(FileUtils.java:781) ~[ModpackDownloader.jar:?]
        at org.apache.commons.io.FileUtils.copyFileToDirectory(FileUtils.java:883) ~[ModpackDownloader.jar:?]
        at org.apache.commons.io.FileUtils.copyFileToDirectory(FileUtils.java:844) ~[ModpackDownloader.jar:?]
        at com.nincraft.modpackdownloader.util.FileSystemHelper.moveFromLocalRepo(FileSystemHelper.java:40) ~[ModpackDownloader.jar:?]
        at com.nincraft.modpackdownloader.util.DownloadHelper.downloadFile(DownloadHelper.java:84) ~[ModpackDownloader.jar:?]
        at com.nincraft.modpackdownloader.util.DownloadHelper.downloadFile(DownloadHelper.java:37) ~[ModpackDownloader.jar:?]
        at com.nincraft.modpackdownloader.handler.CurseFileHandler.downloadCurseMod(CurseFileHandler.java:48) ~[ModpackDownloader.jar:?]
        at com.nincraft.modpackdownloader.handler.CurseFileHandler.downloadMod(CurseFileHandler.java:264) ~[ModpackDownloader.jar:?]
        at com.nincraft.modpackdownloader.processor.DownloadModsProcessor.lambda$downloadMods$1(DownloadModsProcessor.java:49) ~[ModpackDownloader.jar:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
```

I noticed that the jar filename had a space in it, and I suspected some sort of escaping problem. I checked the cache directory, and the file was there, but under a wrong name: `InventoryTweaks-1.64+dev.151.jar` (ie the space had been replaced with a plus). 

Additionally, at the end of the run it was reported

```
19:57:50.121 [main] INFO  com.nincraft.modpackdownloader.summary.DownloadSummarizer:39 - Successfully downloaded 253 mods
```

I'd expect that errors finding the downloaded mods in the cache folder would have caused the download process to fail (and thus break my Docker build).

**Comment by UndeadZeratul** [comment]
What version of MPD are you running with?  I only ask because I've been making changes to align with the newer profile JSON format in v0.7, and depending on the format that the modpack you're trying to build uses, the solution would either need to be a hotfix on the old code or just a fix in the newer code and a new version pushed out.

I've seen the failure to copy from local cache in my local testing as well, so I might as well fix for both of us, but if it's going to be a hotfix on the existing release's code we'll fix it in a 0.6.1 bump.  Also if it's possible, do you have the JSON configs for the modpack so I may try and test it?

**Comment by mythmon** [comment]
I'm using 0.6, but I'd be willing to upgrade to 0.7 if it's not too big a change.

By JSON configs, do you mean the manifest file? I included that in the gist I linked to in my first comment. Here it is again in case something was wrong with the first one 

https://gist.github.com/mythmon/589ea705768b2b004ca856ee321ca6d0#file-manifest-json

**Comment by UndeadZeratul** [comment]
Ah, yes the manifest; I'll be honest I didn't click the link so I apologize.  Thanks for the link again though.  Looking at the file it's definitely the old structure.  Given how many mods you have in the list, I'd say it could be a big change, but if you have the modpack loaded in Twitch and you open the profile folder or export it, does it still give you the `fileID` and `projectID` properties in the `files` list?

The new structure is a bit different.  the `minecraft` config seems to have gone away and instead of `modloaders` there's simply a `baseModLoader` that holds the Forge download info, including the full download URL.  In addition, the `files` list is now `installedAddons`, which contain the mod metadata and the `installedFiles` list for that addon.  That list is usually going to be only one file, but it's still a list MPD has to iterate through to grab each file.  The good part there is the file object has the full download URL, which actually simplified our processors quite a bit.

The downside to the change is that pack updating had to be disabled until we figure out the proper method going forward to update the manifest, although going through your use case I doubt that'll be an issue.  On top of that, we don't have a way to migrate the old manifest to the new, which is why I asked if Twitch can update the manifest for you, take the legwork out of that step.  If/when you can get the manifest updated, then v0.7 should be able to continue working, albeit copying the file would still throw an error.

I have noticed though that once it failed to copy the file, it still downloaded the file as normal successfully; does that not happen in v0.6?  It looks like it does from the log, if it's not then that's still a bug...

```
19:57:36.064 [pool-2-thread-24] ERROR com.nincraft.modpackdownloader.util.FileSystemHelper:48 - Could not copy better-builders-wands from local repo.
java.io.FileNotFoundException: Source '/root/.modpackdownloader/better-builders-wands/BetterBuildersWands-1.12.2-0.13.2.271 5997513.jar' does not exist
        at org.apache.commons.io.FileUtils.checkFileRequirements(FileUtils.java:274) ~[ModpackDownloader.jar:?]
        at org.apache.commons.io.FileUtils.copyFile(FileUtils.java:781) ~[ModpackDownloader.jar:?]
        at org.apache.commons.io.FileUtils.copyFileToDirectory(FileUtils.java:883) ~[ModpackDownloader.jar:?]
        at org.apache.commons.io.FileUtils.copyFileToDirectory(FileUtils.java:844) ~[ModpackDownloader.jar:?]
        at com.nincraft.modpackdownloader.util.FileSystemHelper.moveFromLocalRepo(FileSystemHelper.java:40) ~[ModpackDownloader.jar:?]
        at com.nincraft.modpackdownloader.util.DownloadHelper.downloadFile(DownloadHelper.java:84) ~[ModpackDownloader.jar:?]
        at com.nincraft.modpackdownloader.util.DownloadHelper.downloadFile(DownloadHelper.java:37) ~[ModpackDownloader.jar:?]
        at com.nincraft.modpackdownloader.handler.CurseFileHandler.downloadCurseMod(CurseFileHandler.java:48) ~[ModpackDownloader.jar:?]
        at com.nincraft.modpackdownloader.handler.CurseFileHandler.downloadMod(CurseFileHandler.java:264) ~[ModpackDownloader.jar:?]
        at com.nincraft.modpackdownloader.processor.DownloadModsProcessor.lambda$downloadMods$1(DownloadModsProcessor.java:49) ~[ModpackDownloader.jar:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
19:57:36.167 [pool-2-thread-24] INFO  com.nincraft.modpackdownloader.util.DownloadHelper:85 - Successfully downloaded BetterBuildersWands-1.12.2-0.13.2.271+5997513.jar19:57:36.169 [pool-2-thread-70] INFO  com.nincraft.modpackdownloader.util.DownloadHelper:85 - Successfully downloaded despawningspawners-1.12-1.2.0-2.jar
```

_(looking at that now, looks like the log dropped a newline somewhere, `despawningspawners` logged on the same line as `BetterBuildersWands` being "successfully downloaded".)_

**Comment by UndeadZeratul** [comment]
sigh, forgot to link it in the commit message, but https://github.com/Nincraft/ModPackDownloader/commit/c0f9e1c173528235ef6f33131ae21727a9388e4a should fix the issue.

**Comment by UndeadZeratul** [comment]
Looking it over, it should be straightforward to roll this down to v0.6 and push out a v0.6.1.  I'll try and get it out in a day or so if that's ok.

**Comment by mythmon** [comment]
> if you have the modpack loaded in Twitch and you open the profile folder or export it, does it still give you the fileID and projectID properties in the files list?

I don't use Twitch, so I'm not sure what this is. I generally use MultiMC.

> I have noticed though that once it failed to copy the file, it still downloaded the file as normal successfully; does that not happen in v0.6? It looks like it does from the log, if it's not then that's still a bug...

The file is downloaded to the cache, it just isn't copied to the target directory.

> Looking it over, it should be straightforward to roll this down to v0.6 and push out a v0.6.1. I'll try and get it out in a day or so if that's ok.

Sounds great! I've got a work-around in place for now, so I'm not blocked on anything.

Also, in case it wasn't clear, I'm not a mod author, I'm just trying to use a mod without going through one of the normal launchers. I'm trying to build up a Dockerfile I can use to run a server.

**Comment by UndeadZeratul** [comment]
Here's the new release, by the way: https://github.com/Nincraft/ModPackDownloader/releases/tag/0.6.1

Let me know if that doesn't fix it and I'll reopen the issue (it auto-closed from the commit message)

Oh, also, to respond to your responses:

> I don't use Twitch, so I'm not sure what this is. I generally use MultiMC.

Ah, no worries, I only asked because you had the old manifest, I assumed it was from the Twitch Launcher at some point.  If that profile was still installed you could potentially re-export it and see if it would convert the old manifest to the new, but I haven't tested it so I'm not sure.  We'll be looking into making our JSON accept and import old Twitch manifests but ultimately use our own data structure for developing packs, then export them if Twitch (or Overwolf now, I guess) continues to use it.  Otherwise, the structure won't matter because we'd only be using it to tell MPD where to get files, which it looks like Twitch ended up giving us a direct URL and the widget we use for updating gave us enough information to build that same URL, so assuming that doesn't change, it'll be much more straightforward in v0.7+ once we finalize our own manifest data structure.

> The file is downloaded to the cache, it just isn't copied to the target directory.

Ah cool, that's what I saw, and should be fixed with v0.6.1, I even used your supplied manifest to test and make sure the file was copied over.

> Also, in case it wasn't clear, I'm not a mod author, I'm just trying to use a mod without going through one of the normal launchers. I'm trying to build up a Dockerfile I can use to run a server.

I figured as much, I just figured I'd ask to get information on how to properly fix the issue.  By the way, I will say you're one of the first people to use MPD that seems to have actually used it for CI/CD purposes while working on a modpack.  Not that you are a SevTech: Ages pack dev, but like we've talked about toying with Docker for deploying a server for our own modpack projects back in 1.7.10.  I'm glad to see more people looking into the development workflow for modpacks, not just mods.
